File: test/TestCases/node-addon-api/bigint.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

'use strict';
const assert = require('assert');

module.exports = require('../common').runTest(test);

function test (binding) {
  const {
    isLossless,
    isBigInt,
    testInt64,
    testUInt64,
    testWords,
    testWordSpan,
    testBigInteger
  } = binding.bigInt;

  [
    0n,
    -0n,
    1n,
    -1n,
    100n,
    2121n,
    -1233n,
    986583n,
    -976675n,
    98765432213456789876546896323445679887645323232436587988766545658n,
    -4350987086545760976737453646576078997096876957864353245245769809n
  ].forEach((num) => {
    if (num > -(2n ** 63n) && num < 2n ** 63n) {
      assert.strictEqual(testInt64(num), num);
      assert.strictEqual(isLossless(num, true), true);
    } else {
      assert.strictEqual(isLossless(num, true), false);
    }

    if (num >= 0 && num < 2n ** 64n) {
      assert.strictEqual(testUInt64(num), num);
      assert.strictEqual(isLossless(num, false), true);
    } else {
      assert.strictEqual(isLossless(num, false), false);
    }

    assert.strictEqual(isBigInt(num), true);

    assert.strictEqual(num, testWords(num));
    assert.strictEqual(num, testWordSpan(num));
    assert.strictEqual(num, testBigInteger(num, num.toString()));
  });
}

File: test/TestCases/node-addon-api/bigint.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System;
using System.Numerics;
using Microsoft.JavaScript.NodeApi;

namespace Microsoft.JavaScript.NodeApiTest;

public class TestBigInt : TestHelper, ITestObject
{
    private static JSValue IsLossless(JSCallbackArgs args)
    {
        JSBigInt big = (JSBigInt)args[0];
        bool isSigned = (bool)args[1];

        if (isSigned)
        {
            big.ToInt64(out bool isLossless);
            return isLossless;
        }
        else
        {
            big.ToUInt64(out bool isLossless);
            return isLossless;
        }
    }

    private static JSValue IsBigInt(JSCallbackArgs args)
    {
        JSBigInt? big = args[0].AsJSBigInt();
        return big.HasValue;
    }

    private static JSValue TestInt64(JSCallbackArgs args)
    {
        JSBigInt big = (JSBigInt)args[0];
        long input = big.ToInt64(out _);
        return new JSBigInt(input);
    }

    private static JSValue TestUInt64(JSCallbackArgs args)
    {
        JSBigInt big = (JSBigInt)args[0];
        ulong input = big.ToUInt64(out _);
        return new JSBigInt(input);
    }

    private static JSValue TestWords(JSCallbackArgs args)
    {
        JSBigInt big = (JSBigInt)args[0];
        ulong[] words = big.ToUInt64Array(out int sign);
        return new JSBigInt(sign, words);
    }

    private static JSValue TestWordSpan(JSCallbackArgs args)
    {
        JSBigInt big = (JSBigInt)args[0];
        int expectedWordCount = big.GetWordCount();
        Span<ulong> words = stackalloc ulong[10];
        big.CopyTo(words, out int sign, out int wordCount);

        if (wordCount != expectedWordCount)
        {
            JSError.ThrowError("word count did not match");
            return default;
        }

        return new JSBigInt(sign, words.Slice(0, wordCount));
    }

    private static JSValue TestBigInteger(JSCallbackArgs args)
    {
        JSBigInt big = (JSBigInt)args[0];
        string bigStr = (string)args[1];
        BigInteger expected = BigInteger.Parse(bigStr);
        BigInteger actual = big.ToBigInteger();
        if (expected != actual)
        {
            JSError.ThrowError("BigInteger is not parsed correctly");
            return default;
        }

        return new JSBigInt(actual);
    }

    public JSObject Init() => [
        Method(IsLossless),
        Method(IsBigInt),
        Method(TestInt64),
        Method(TestUInt64),
        Method(TestWords),
        Method(TestWordSpan),
        Method(TestBigInteger),
    ];
}
